packages/zoe/test/unitTests/contracts/test-autoswap.js

```diff
@@ -13,8 +14,348 @@ import { assertOfferResult, assertPayoutAmount } from '../../zoeTestHelpers';
+      } else {
+        t.deepEquals(moola(mPost), poolPost.Secondary, `moola after swap`);
+        t.deepEquals(simoleans(sPost), poolPost.Central, `s after swap`);
+      }
+      t.equals(kPost, sPost * mPost);
+
+      seat
+        .getOfferResult()
+        .then(async () =>
+          t.equals(
+            lPost,
+            await E(publicFacet).getLiquidityTokens(),
+            'liquidity after',
+          ),
+        );
+    },
+
+    addLiquidityAndCheck: async (t, priorPoolState, details, expected) => {
+      // just check that it went through, and the results are as stated.
+      // The test will declare fees, refunds, and figure out when the trade
+      // gets less than requested
+      const { m: mPre, s: sPre, l: lPre, k: kPre } = priorPoolState;
+      const { mAmount, sAmount, lAmount = 0 } = details;
+      const {
+        m: mPost,
+        s: sPost,
+        l: lPost,
+        k: kPost,
+        payoutL,
+        payoutM,
+        payoutS,
+      } = expected;
+      t.assert(payoutM === 0 || payoutS === 0, 'only refund one side');
+      const mIsBasis =
+        multiply(mAmount.value, sPre) <= multiply(sAmount.value, mPre);
+      const scaleByAlpha = mIsBasis
+        ? makeScaleFn(mPre, mPost)
+        : makeScaleFn(sPre, sPost);
+
+      const poolPre = await E(publicFacet).getPoolAllocation();
+      t.deepEquals(moola(mPre), poolPre.Central, `moola before add liq`);
+      t.deepEquals(simoleans(sPre), poolPre.Secondary, `s before add liq`);
+      t.equals(
+        lPre,
+        await E(publicFacet).getLiquidityTokens(),
+        'liquidity pool before add',
+      );
+      t.equals(kPre, sPre * mPre);
+
+      const proposal = harden({
+        give: { Central: mAmount, Secondary: sAmount },
+        want: { Liquidity: liquidity(lAmount) },
+      });
+      const payment = harden({
+        Central: withdrawPayment(mAmount),
+        Secondary: withdrawPayment(sAmount),
+      });
+
+      const seat = await zoe.offer(
+        E(publicFacet).makeAddLiquidityInvitation(),
+        proposal,
+        payment,
+      );
+      assertOfferResult(t, seat, 'Added liquidity.');
+
+      const {
+        Central: mPayout,
+        Secondary: sPayout,
+        Liquidity: lPayout,
+      } = await seat.getPayouts();
+      assertPayoutAmount(t, moolaIssuer, mPayout, moola(payoutM));
+      assertPayoutAmount(t, simoleanIssuer, sPayout, simoleans(payoutS));
+      assertPayoutAmount(t, liquidityIssuer, lPayout, liquidity(payoutL));
+
+      const poolPost = await E(publicFacet).getPoolAllocation();
+      t.deepEquals(moola(mPost), poolPost.Central, `moola after add liq`);
+      t.deepEquals(simoleans(sPost), poolPost.Secondary, `s after add liq`);
+      t.equals(
+        lPost,
+        await E(publicFacet).getLiquidityTokens(),
+        'liquidity pool after',
+      );
+      t.equals(kPost, sPost * mPost, 'expected value of K after addLiquidity');
```

Comment:
hmm, isn't the constant product such that we shouldn't have a `kPost` variable? Constant product would mean that we could use `kPre` and assert that `sPost * mPost` >= than `kPre`.

Reply:
Ah, gotcha. I forgot that it only applies to swaps. Adding or removing liquidity would affect k considerably. Looks good tome!